doc/greydb.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "GREYDB" "8" "December 2019" "" ""
.
.SH "NAME"
\fBgreydb\fR \- greyd database tool
.
.SH "SYNOPSIS"
\fBgreydb\fR [\fB\-f\fR config] [[\fB\-TDt\fR] \fB\-a\fR keys] [[\fB\-TDt\fR] \fB\-d\fR keys] [\fB\-Y\fR synctarget]
.
.SH "DESCRIPTION"
\fBgreydb\fR manipulates the \fBgreyd\fR database used for \fBgreyd\fR(8)\.
.
.P
The options are as follows:
.
.TP
\fB\-f\fR \fIconfig\fR
The main greyd configuration file\.
.
.TP
\fB\-a\fR \fIkeys\fR
Add or update the entries for keys\. This can be used to whitelist one or more IP addresses (i\.e\. circumvent the greylisting process altogether) by adding all IP addresses as keys to the \fBgreyd\fR database for WHITE entries\. If any keys specified match entries already in the \fBgreyd\fR database, \fBgreydb\fR updates the entry\'s time last seen to now\.
.
.TP
\fB\-d\fR \fIkeys\fR
Delete entries for keys\.
.
.TP
\fB\-T\fR
Add or delete the keys as SPAMTRAP entries\. See the GREYTRAPPING section of \fBgreyd\fR(8) for more information\. Must be used in conjunction with the \fB\-a\fR or \fB\-d\fR option\.
.
.TP
\fB\-D\fR
Add or delete the keys as permitted DOMAIN entries\. See the GREYTRAPPING section of \fBgreyd\fR(8) for more information\. Must be used in conjunction with the \fB\-a\fR or \fB\-d\fR option\.
.
.TP
\fB\-t\fR
Add or delete the keys as TRAPPED entries\. See the GREYTRAPPING section of \fBgreyd\fR(8) for more information\. Must be used in conjunction with the \fB\-a\fR or \fB\-d\fR option\.
.
.TP
\fB\-Y\fR \fIsynctarget\fR
Add a target to receive synchronisation messages; see \fISYNCHRONISATION\fR below\. This option can be specified multiple times\.
.
.P
If adding or deleting a SPAMTRAP address (\fB\-T\fR), keys should be specified as email addresses:
.
.IP "" 4
.
.nf

spamtrap@mydomain\.org
.
.fi
.
.IP "" 0
.
.P
If adding or deleting a DOMAIN entries (\fB\-D\fR), keys should be specified as domains/email addresses:
.
.IP "" 4
.
.nf

allowed\-domain\.org
@another\-allowed\-domain\.org
individual@greyd\.org
.
.fi
.
.IP "" 0
.
.P
Otherwise keys must be numerical IP addresses\.
.
.SH "DATABASE OUTPUT FORMAT"
If invoked without any arguments, \fBgreydb\fR lists the contents of the database in a text format\.
.
.P
For SPAMTRAP and DOMAIN entries the format is:
.
.IP "" 4
.
.nf

type|mailaddress
.
.fi
.
.IP "" 0
.
.P
where type will be SPAMTRAP and mailaddress will be the email address for which any connections received by \fBgreyd\fR(8) will be blacklisted if mail is sent to this address\.
.
.P
For TRAPPED entries the format is:
.
.IP "" 4
.
.nf

type|ip|expire
.
.fi
.
.IP "" 0
.
.P
where \fItype\fR will be TRAPPED, IP will be the IP address blacklisted due to hitting a spamtrap, and \fIexpire\fR will be when the IP is due to be removed from the blacklist\.
.
.P
For GREY entries, the format is:
.
.IP "" 4
.
.nf

type|source IP|helo|from|to|first|pass|expire|block|pass
.
.fi
.
.IP "" 0
.
.P
For WHITE entries, the format is:
.
.IP "" 4
.
.nf

type|source IP|||first|pass|expire|block|pass
.
.fi
.
.IP "" 0
.
.P
The fields are as follows:
.
.TP
\fBtype\fR
WHITE if whitelisted or GREY if greylisted
.
.TP
\fBsource IP\fR
IP address the connection originated from
.
.TP
\fBhelo\fR
what the connecting host sent as identification in the HELO/EHLO command in the SMTP dialogue
.
.TP
\fBfrom\fR
envelope\-from address for GREY (empty for WHITE entries)
.
.TP
\fBto\fR
envelope\-to address for GREY (empty for WHITE entries)
.
.TP
\fBfirst\fR
time the entry was first seen
.
.TP
\fBpass\fR
time the entry passed from being GREY to being WHITE
.
.TP
\fBexpire\fR
time the entry will expire and be removed from the database
.
.TP
\fBblock\fR
number of times a corresponding connection received a temporary failure from \fBgreyd\fR(8)
.
.TP
\fBpass\fR
number of times a corresponding connection has been seen to pass to the real MTA by \fBgreylogd\fR(8)
.
.P
Note that times are in seconds since the Epoch, in the manner returned by time(3)\. Times may be converted to human readable format using:
.
.IP "" 4
.
.nf

$ date \-\-date \'@<value>\'
.
.fi
.
.IP "" 0
.
.SH "SYNCHRONISATION"
\fBgreydb\fR supports realtime synchronisation of added entries by sending the information it updates to a number of \fBgreyd\fR(8) daemons running on multiple machines\. To enable synchronisation, use the command line option \-Y to specify the machines to which \fBgreydb\fR will send messages\. The synchronisation may also be configured entirely via \fBgreyd\.conf\fR(5)\. For more information, see \fBgreyd\fR(8) and \fBgreyd\.conf\fR(5)\.
.
.P
\fBgreydb\fR only sends sync messages for additions/deletions of WHITE & TRAPPED entries only\.
.
.SH "COPYRIGHT"
\fBgreydb\fR is Copyright (C) 2015 Mikey Austin (greyd\.org)
.
.SH "SEE ALSO"
\fBgreyd\.conf\fR(5), \fBgreyd\-setup\fR(8), \fBgreyd\fR(8), \fBgreylogd\fR(8)
.
.SH "CREDITS"
Much of this man page was taken from the \fIOpenBSD\fR manual, and adapted accordingly\.


NEWS
2015-07-15
        greyd now makes use of libev for the main event loop & connection
        handling.

2015-06-08
        greyd now sports sync compatibility with spamd.

2015-04-15
        A host of new database drivers are now available to support different
        use cases & workloads. The blacklist storage is now implemented by
        way of a PATRICIA trie, which makes lookups 1000x faster than previous
        linear search.

2015-02-01
        Added new SPF validation functionality.

2015-01-28
        DragonFly BSD is now supported using the pf firewall driver.

2015-01-24
        NetBSD is now supported using the pf firewall driver. Syncing works between
        any host running greyd.

2015-01-19
        There has been an internal re-shuffle in order to isolate the parts of the
        system that hold a firewall handle to one unprivileged process, with the main
        connection handling process & grey processes communicating with it over a pipe.
        Now the main connection handling process has no privileges, and runs chrooted
        with no direct access to the database (may run as a different user) nor the
        firewall.

2015-01-07
        Greyd now has a new PF firewall driver, and will be included in release 0.2.0.

src/greydb.c
/*
 * Copyright (c) 2014, 2015 Mikey Austin <mikey@greyd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/**
 * @file   greydb.c
 * @brief  Implements the generic greylisting database abstraction.
 * @author Mikey Austin
 * @date   2014
 */

#include "greydb.h"
#include "constants.h"
#include "failures.h"
#include "mod.h"

#include <stdlib.h>

extern DB_handle_T
DB_init(Config_T config)
{
    DB_handle_T handle;
    Config_section_T section;
    char* user;

    /* Setup the db handle. */
    if ((handle = malloc(sizeof(*handle))) == NULL) {
        i_critical("Could not create db handle");
    }

    if ((section = Config_get_section(config, "database")) == NULL) {
        i_critical("Could not find database configuration");
    }

    handle->config = config;
    if (Config_get_int(config, "drop_privs", NULL, 1)
        && (user = Config_get_str(config, "user", "grey", GREYD_DB_USER)) != NULL) {
        if ((handle->pw = getpwnam(user)) == NULL) {
            i_critical("No such user %s", user);
        }
    } else {
        handle->pw = NULL;
    }

    /* Open the configured driver and extract all required symbols. */
    handle->driver = Mod_open(section, "db");

    handle->db_init = (void (*)(DB_handle_T))
        Mod_get(handle->driver, "Mod_db_init");
    handle->db_open = (void (*)(DB_handle_T, int))
        Mod_get(handle->driver, "Mod_db_open");
    handle->db_start_txn = (int (*)(DB_handle_T))
        Mod_get(handle->driver, "Mod_db_start_txn");
    handle->db_commit_txn = (int (*)(DB_handle_T))
        Mod_get(handle->driver, "Mod_db_commit_txn");
    handle->db_rollback_txn = (int (*)(DB_handle_T))
        Mod_get(handle->driver, "Mod_db_rollback_txn");
    handle->db_close = (void (*)(DB_handle_T))
        Mod_get(handle->driver, "Mod_db_close");
    handle->db_put = (int (*)(DB_handle_T, struct DB_key*, struct DB_val*))
        Mod_get(handle->driver, "Mod_db_put");
    handle->db_get = (int (*)(DB_handle_T, struct DB_key*, struct DB_val*))
        Mod_get(handle->driver, "Mod_db_get");
    handle->db_del = (int (*)(DB_handle_T, struct DB_key*))
        Mod_get(handle->driver, "Mod_db_del");
    handle->db_get_itr = (void (*)(DB_itr_T, int))
        Mod_get(handle->driver, "Mod_db_get_itr");
    handle->db_itr_next = (int (*)(DB_itr_T, struct DB_key*, struct DB_val*))
        Mod_get(handle->driver, "Mod_db_itr_next");
    handle->db_itr_replace_curr = (int (*)(DB_itr_T, struct DB_val*))
        Mod_get(handle->driver, "Mod_db_itr_replace_curr");
    handle->db_itr_del_curr = (int (*)(DB_itr_T))
        Mod_get(handle->driver, "Mod_db_itr_del_curr");
    handle->db_itr_close = (void (*)(DB_itr_T))
        Mod_get(handle->driver, "Mod_db_itr_close");
    handle->db_scan = (int (*)(DB_handle_T, time_t*, List_T, List_T,
        List_T, time_t*))
        Mod_get(handle->driver, "Mod_scan_db");

    /* Initialize the database driver. */
    handle->db_init(handle);

    return handle;
}

extern void
DB_open(DB_handle_T handle, int flags)
{
    handle->db_open(handle, flags);
}

extern int
DB_start_txn(DB_handle_T handle)
{
    return handle->db_start_txn(handle);
}

extern int
DB_commit_txn(DB_handle_T handle)
{
    return handle->db_commit_txn(handle);
}

extern int
DB_rollback_txn(DB_handle_T handle)
{
    return handle->db_rollback_txn(handle);
}

extern void
DB_close(DB_handle_T* handle)
{
    if (handle == NULL || *handle == NULL) {
        return;
    }

    (*handle)->db_close(*handle);
    Mod_close((*handle)->driver);
    free(*handle);
    *handle = NULL;
}

extern int
DB_put(DB_handle_T handle, struct DB_key* key, struct DB_val* val)
{
    return handle->db_put(handle, key, val);
}

extern int
DB_get(DB_handle_T handle, struct DB_key* key, struct DB_val* val)
{
    return handle->db_get(handle, key, val);
}

extern int
DB_del(DB_handle_T handle, struct DB_key* key)
{
    return handle->db_del(handle, key);
}

extern DB_itr_T
DB_get_itr(DB_handle_T handle, int types)
{
    DB_itr_T itr;

    /* Setup the iterator. */
    if ((itr = malloc(sizeof(*itr))) == NULL) {
        i_critical("Could not create iterator");
    }

    itr->handle = handle;
    itr->current = -1;
    itr->size = 0;

    handle->db_get_itr(itr, types);

    return itr;
}

extern int
DB_itr_next(DB_itr_T itr, struct DB_key* key, struct DB_val* val)
{
    return itr->handle->db_itr_next(itr, key, val);
}

extern int
DB_itr_replace_curr(DB_itr_T itr, struct DB_val* val)
{
    return itr->handle->db_itr_replace_curr(itr, val);
}

extern int
DB_itr_del_curr(DB_itr_T itr)
{
    return itr->handle->db_itr_del_curr(itr);
}

extern void
DB_close_itr(DB_itr_T* itr)
{
    if (itr == NULL || *itr == NULL) {
        return;
    }

    (*itr)->handle->db_itr_close(*itr);
    free(*itr);
    *itr = NULL;
}

extern int
DB_scan(DB_handle_T handle, time_t* now, List_T whitelist,
    List_T whitelist_ipv6, List_T traplist, time_t* white_exp)
{
    return handle->db_scan(handle, now, whitelist, whitelist_ipv6, traplist,
        white_exp);
}

extern int
DB_addr_state(DB_handle_T handle, char* addr)
{
    struct DB_key key;
    struct DB_val val;
    struct Grey_data gd;
    int ret;

    key.type = DB_KEY_IP;
    key.data.s = addr;
    ret = DB_get(handle, &key, &val);

    switch (ret) {
    case GREYDB_NOT_FOUND:
        return 0;

    case GREYDB_FOUND:
        gd = val.data.gd;
        return (gd.pcount == -1 ? 1 : 2);

    default:
        return -1;
    }
}
